File: 6주차/얼려먹기/DFS_ex02.py
# 음료수 얼려먹기 문제

n, m = map(int, input().split())

graph = []
for i in range(n):
    graph.append(list(map(int, input())))


def dfs(x, y):
    # 주어진 범위를 벗어나는 경우 즉시 종료
    if x <= -1 or x >= n or y <= -1 or y >= m:
        return False
    
    # 현재 노드를 아직 방문하지 않았다면
    if graph[x][y] == 0:
        # 방문 처리
        graph[x][y] = 1
        # 상하좌우의 위치들을 모두 재귀적으로 호출
        dfs(x -1 , y)
        dfs(x, y -1)
        dfs(x + 1, y)
        dfs(x, y +1)
        return True
    return False


# 모든 노드에 대하여 탐색한 후 0을 다 탐색하면 result + 1
result = 0
for i in range(n):
    for j in range(m):
        if dfs(i, j) == True:
            result += 1

print(result)

File: 6주차/얼려먹기/DFS_ex02.java
package 얼려먹기;
// 음료수 얼려먹기 문제

import java.util.Arrays;
import java.util.Scanner;

public class DFS_ex02 {
    public static int N, M;
 
    public static boolean dfs(int[][] graph, int x, int y) {
        if(x <= 1 || x >= N || y <= -1 || y >= M)
            return false;

        if(graph[x][y] == 0)
        {
            graph[x][y] = 1;
            dfs(graph, x -1, y);
            dfs(graph, x, y -1);
            dfs(graph, x +1, y);
            dfs(graph, x, y +1);
            return true;
        }

        return false;
    }

    public static void main(String[] args) 
    {
        Scanner scanner = new Scanner(System.in);
        // N, M 입력
        N = scanner.nextInt();
        M = scanner.nextInt();
        scanner.nextLine();
        System.out.println("+++++++++");
        System.out.println("n: " + N + " m: " + M);
        int [][] graph = new int[N][M];

        // 2차원 리스트의 맵 정보 입력 받기
        for(int i=0; i<N; i++)
        {
            String str = scanner.nextLine();
            for(int j=0; j<M; j++){
                graph[i][j] = str.charAt(j) - '0';
            }
        }
        scanner.close();
        for(int i=0; i<N; i++){
            System.out.println(Arrays.toString(graph[i]));
        }


        int result = 0;
        // 2차원 리스트의 맵 정보 입력 받기
        for(int i=0; i<N; i++)
        {
            for(int j=0; j<M; j++)
            {
                if(dfs(graph, i, j))
                    result += 1;
            }
        }
        System.out.println(result);
    }
}
